packages/core/src/types.ts

```diff
@@ -423,55 +440,105 @@ export type TransitionsConfig<
     : TransitionConfigOrTarget<TContext, ExtractEvent<TEvent, K>, TEvent>;
 };
 
-export interface InvokeConfig<
+export type InvokeConfig<
   TContext extends MachineContext,
-  TEvent extends EventObject
-> {
-  /**
-   * The unique identifier for the invoked machine. If not specified, this
-   * will be the machine's own `id`, or the URL (from `src`).
-   */
-  id?: string;
-
-  systemId?: string;
-  /**
-   * The source of the machine to be invoked, or the machine itself.
-   */
-  src: string | ActorLogic<any, any>; // TODO: fix types
-
-  input?: Mapper<TContext, TEvent, any> | any;
-  /**
-   * The transition to take upon the invoked child machine reaching its final top-level state.
-   */
-  onDone?:
-    | string
-    | SingleOrArray<
-        TransitionConfigOrTarget<TContext, DoneInvokeEvent<any>, TEvent>
-      >;
-  /**
-   * The transition to take upon the invoked child machine sending an error event.
-   */
-  onError?:
-    | string
-    | SingleOrArray<
-        TransitionConfigOrTarget<TContext, ErrorEvent<any>, TEvent>
-      >;
-
-  onSnapshot?:
-    | string
-    | SingleOrArray<
-        TransitionConfigOrTarget<TContext, SnapshotEvent<any>, TEvent>
-      >;
-  /**
-   * Meta data related to this invocation
-   */
-  meta?: MetaObject;
-}
+  TEvent extends EventObject,
+  TActor extends ProvidedActor
+> = TActor extends { src: infer TSrc }
+  ? {
+      /**
+       * The unique identifier for the invoked machine. If not specified, this
+       * will be the machine's own `id`, or the URL (from `src`).
+       */
+      id?: TActor['id'];
```

Comment:
hm, when the `id` is provided I feel like it's better to require, thoughts?

Reply:
You mean when it's provided to the `{ actors: ... }` type?

Reply:
Yeah, if any of the given `src` has an associated `id` with it then it feels like we should *require* it here.

Reply:
Okay so to clarify:
```ts
createMachine({
  type: {} as {
    actors: {
      src: 'someSrc';
      logic: SomeLogic;
      id: 'foo' | 'bar';
    }
  },
  invoke: {
    src: 'someSrc',
    // no id = type error
  }
})
```

I agree, if that makes things simpler

Reply:
done in https://github.com/statelyai/xstate/pull/4036/commits/68609041e469e239fac67874eb27cef5ab024901